### PR TITLE
CRM: Add backend action Invoice Set Status and Invoice Event Manager

### PR DESCRIPTION
**Pull request by gogdzl**
This PR adds the action Invoice Set Status, and the needed event manager for invoices.

## Proposed changes:
* Add Invoice Set Status Action
* Add Invoice Event Manager

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
https://github.com/Automattic/zero-bs-crm/issues/3331
https://github.com/Automattic/zero-bs-crm/issues/3195

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:

* Check the code
* Check the coverage of our new tests
* Run the tests and make sure they are passing

### CODE REVIEW

File: projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php [hunk 1]
```diff
@@ -1900,6 +1908,7 @@ public function addUpdateInvoice($args=array()){
                             'extraMeta'=>$confirmedExtraMeta #} This is the "extraMeta" passed (as saved)
                         ));
 
+												$this->events_manager->invoice()->created( array_merge( $data, array( 'id' => $newID ) ) ); // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase
```
Comment (cleacos) [inline projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php]: Why not: `$data['id'] = $newID` before launching the event? and pass the `$data` without creating an array copy.

File: projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+	/**
+	 * Get the title of the step.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return string|null The title of the step.
+	 */
+	public static function get_title(): ?string {
+		return 'Set Invoice Status Action';
```
Comment (cleacos) [inline projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php]: Translation.

File: projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+	/**
+	 * Get the description of the step.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return string|null The description of the step.
+	 */
+	public static function get_description(): ?string {
+		return 'Action to set the invoice status';
```
Comment (cleacos) [inline projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php]: Translation.

File: projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php (hunk 1 above)
Comment (gogdzl) [reply projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php]: I was just being extra careful not to change variables that could be used later, but `$data` is not used anywhere else, so I agree that it is safe to change it. Addressed in da42ac3.

File: projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php (hunk 3 above)
Comment (gogdzl) [reply projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php]: 8403a19

File: projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php (hunk 2 above)
Comment (gogdzl) [reply projects/plugins/crm/src/automation/commons/actions/invoice/class-set-invoice-status.php]: 8403a19

File: projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php [hunk 4]
```diff
@@ -1741,7 +1749,7 @@ public function addUpdateInvoice($args=array()){
 									'prev_invoice'   => $previous_invoice_obj,
                                     ));
 
-                                
+											$this->events_manager->invoice()->updated( $data );
```
Comment (cleacos) [inline projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php:1753]: Here too, we have to add the ID info to `$data`

`$data['id'] = $id;`

**Review by cleacos — DISMISSED** [review]
The code looks good. Tests pass.

LGTM!

**Review by cleacos — APPROVED** [review]
Cool! Now tests pass again.

LGTM!

**Pull request merged**